### ISSUE TITLE
ESPHome devices not configurable as "Individual devices" in Home Assistant's Energy dashboard

### ISSUE THREAD

**Issue by alexiri** [esphome/issues#3661]
### The problem

Sensors with the `total_daily_energy` platform are not selectable as "Individual devices" in Home Assistant's Energy dashboard. I read the [long-term statistics doc](https://developers.home-assistant.io/docs/core/entity/sensor/#long-term-statistics) and I think my esphome configuration is correct, but I might be missing some crucial detail. 

`sensor.shellyem_channel_1_energy` is a Shelly EM that uses the stock Shelly firmware and integration, and that one works perfectly with the Energy dashboard. `sensor.shelly_1pm_1_daily_energy` is a Shelly 1PM that uses Esphome (configured basically [like this](https://www.esphome-devices.com/devices/Shelly-1PM) but with the `total_daily_energy` snippet listed below), but that one does not show up in the dropdown to configure it as an Individual device.

![image](https://user-images.githubusercontent.com/5055697/194495015-7561f28c-028c-47db-979d-8001d6378a7e.png)

What magic is missing?

### Which version of ESPHome has the issue?

2022.9.4

### What type of installation are you using?

Docker

### Which version of Home Assistant has the issue?

2022.10.1

### What platform are you using?

ESP8266

### Board

shelly

### Component causing the issue

total_daily_energy

### Example YAML snippet

```yaml
- platform: total_daily_energy
    name: "${upper_devicename} Daily Energy"
    device_class: energy
    state_class: total_increasing
    power_id: power
    filters:
      # Multiplication factor from W to kW is 0.001
      - multiply: 0.001
      - lambda: if (x > 0) { return x; } else { return 0; }
    unit_of_measurement: kWh
```

**Comment by alexiri** [comment]
Hi @emontnemery, @bramkragten, do you have any idea why this might be happening? What's missing in ESPHome so that this works?

**Comment by alexiri** [comment]
While digging deeper into the code, I noticed that the missing entities have the old unit of measurement of `W` in the recorder's MySQL database.

![image](https://user-images.githubusercontent.com/5055697/197383944-43398955-8ff6-4930-aeba-d85dabc3a6ec.png)

I manually changed that to `kWh` and now those entities are selectable as Individual Devices! So I guess the question is, why wasn't this column updated in the recorder database? I guess this is no longer an ESPHome issue...
